### PR TITLE
fix(router): Add additional card info in payment response

### PR DESCRIPTION
**Pull request by sahkal**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Currently the payment_method_data is being populated in the payment response is equal to what ever being passed in the request. With this PR we will be able to fetch more card info such as "card_type", "card_issuer" etc in payment response

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Screen Shot 2023-07-20 at 9 04 59 PM
Screen Shot 2023-07-20 at 9 05 51 PM
Screen Shot 2023-07-20 at 9 07 33 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/api_models/src/payments.rs [hunk 1]
```diff
@@ -1125,6 +1137,7 @@ pub enum PaymentMethodDataResponse {
     MandatePayment,
     Reward(RewardData),
     Upi(UpiData),
+    Default,
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:1140]: Remove this variant

Comment (sahkal) [reply crates/api_models/src/payments.rs:1140]: sure

File: crates/api_models/src/payments.rs [hunk 2]
```diff
@@ -1110,21 +1125,21 @@ pub struct RewardData {
...
-    Crypto(CryptoData),
-    BankDebit(BankDebitData),
+    BankRedirect(),
+    Crypto(),
+    BankDebit(),
     MandatePayment,
-    Reward(RewardData),
-    Upi(UpiData),
+    Reward(),
+    Upi(),
```
Comment (Narayanbhat166) [inline crates/api_models/src/payments.rs:1142]: you can remove the brackers

File: crates/router/src/core/payments/helpers.rs [hunk 3]
```diff
@@ -2496,19 +2496,29 @@ pub async fn get_additional_payment_data(
...
+                    api_models::payments::AdditionalCardInfo {
+                        card_issuer: card_data.card_issuer.to_owned(),
+                        card_network: card_data.card_network.clone(),
+                        card_type: card_data.card_type.to_owned(),
+                        card_issuing_country: card_data.card_issuing_country.to_owned(),
+                        bank_code: card_data.bank_code.to_owned(),
+                        last4: card_data.card_number.peek().clone()
+                            [card_number_length - 4..card_number_length]
+                            .to_string(),
+                        card_isin: card_data.card_number.peek().clone()[..6].to_string(),
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/helpers.rs:2516]:
```suggestion
                        card_isin: card_number.get_card_isin()
```

File: crates/router/src/core/payments/helpers.rs [hunk 4]
```diff
@@ -2496,19 +2496,29 @@ pub async fn get_additional_payment_data(
...
-                }
+                api_models::payments::AdditionalPaymentData::Card(Box::new(
+                    api_models::payments::AdditionalCardInfo {
+                        card_issuer: card_data.card_issuer.to_owned(),
+                        card_network: card_data.card_network.clone(),
+                        card_type: card_data.card_type.to_owned(),
+                        card_issuing_country: card_data.card_issuing_country.to_owned(),
+                        bank_code: card_data.bank_code.to_owned(),
+                        last4: card_data.card_number.peek().clone()
+                            [card_number_length - 4..card_number_length]
```
Comment (Narayanbhat166) [inline crates/router/src/core/payments/helpers.rs:2514]: declare a function for card number struct and use it here. And do not use indexing, implement it the same way as get_card_isin() is implemented

**Review by Narayanbhat166 — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments/helpers.rs [hunk 5]
```diff
@@ -2517,22 +2527,40 @@ pub async fn get_additional_payment_data(
...
+                            api_models::payments::AdditionalCardInfo {
+                                card_issuer: card_info.card_issuer,
+                                card_network: card_info.card_network.clone(),
+                                bank_code: card_info.bank_code,
+                                card_type: card_info.card_type,
+                                card_issuing_country: card_info.card_issuing_country,
+                                last4: card_data.card_number.peek().clone()
+                                    [card_number_length - 4..card_number_length]
+                                    .to_string(),
+                                card_isin: card_data.card_number.peek().clone()[..6].to_string(),
```
Comment (SanchithHegde) [inline crates/router/src/core/payments/helpers.rs:2541]: Avoid indexing as it can cause panics, incorporate the changes suggested by Narayan.

File: crates/router/src/core/payments/helpers.rs [hunk 6]
```diff
@@ -2517,22 +2527,40 @@ pub async fn get_additional_payment_data(
...
+                    api_models::payments::AdditionalCardInfo {
+                        card_issuer: None,
+                        card_network: None,
+                        bank_code: None,
+                        card_type: None,
+                        card_issuing_country: None,
+                        last4: card_data.card_number.peek().clone()
+                            [card_number_length - 4..card_number_length]
+                            .to_string(),
+                        card_isin: card_data.card_number.peek().clone()[..6].to_string(),
```
Comment (SanchithHegde) [inline crates/router/src/core/payments/helpers.rs:2558]: Same here

File: crates/api_models/src/payments.rs (hunk 2 above)
Comment (sahkal) [reply crates/api_models/src/payments.rs:1142]: done

File: crates/router/src/core/payments/helpers.rs (hunk 3 above)
Comment (sahkal) [reply crates/router/src/core/payments/helpers.rs:2516]: done

File: crates/router/src/core/payments/helpers.rs (hunk 4 above)
Comment (sahkal) [reply crates/router/src/core/payments/helpers.rs:2514]: done

File: crates/api_models/src/payments.rs [hunk 7]
```diff
@@ -1096,11 +1105,17 @@ pub struct ApplepayPaymentMethod {
...
-    exp_year: String,
+    pub last4: String,
+    pub card_type: Option<String>,
+    pub card_network: Option<api_enums::CardNetwork>,
+    pub card_issuer: Option<String>,
+    pub card_issuing_country: Option<String>,
+    pub card_isin: String,
+    pub card_exp_month: String,
+    pub card_exp_year: String,
+    pub card_holder_name: String,
```
[the remark this answers is not in the document]
Comment (sahkal) [reply crates/api_models/src/payments.rs:1118]: done

File: crates/router/src/core/payments/helpers.rs (hunk 5 above)
Comment (sahkal) [reply crates/router/src/core/payments/helpers.rs:2541]: done

File: crates/router/src/core/payments/helpers.rs (hunk 6 above)
Comment (sahkal) [reply crates/router/src/core/payments/helpers.rs:2558]: +1

File: crates/cards/src/validate.rs [hunk 8]
```diff
@@ -22,6 +22,17 @@ impl CardNumber {
...
+    pub fn get_last4(self) -> String {
+        self.0
+            .peek()
+            .chars()
+            .rev()
+            .take(4)
+            .collect::<String>()
+            .chars()
+            .rev()
+            .collect::<String>()
```
Comment (SanchithHegde) [inline crates/cards/src/validate.rs:34]: Do you need the intermediate `collect()`? Can't you just reverse the iterator and collect it to `String` only once?

```suggestion
        self.0
            .peek()
            .chars()
            .rev()
            .take(4)
            .rev()
            .collect::<String>()
```

File: crates/api_models/src/payments.rs [hunk 9]
```diff
@@ -1797,15 +1797,15 @@ impl From<AdditionalPaymentData> for PaymentMethodDataResponse {
...
         match payment_method_data {
             AdditionalPaymentData::Card(card) => Self::Card(CardResponse::from(*card)),
-            AdditionalPaymentData::PayLater {} => Self::PayLater(),
-            AdditionalPaymentData::Wallet {} => Self::Wallet(),
-            AdditionalPaymentData::BankRedirect { bank_name: _ } => Self::BankRedirect(),
-            AdditionalPaymentData::Crypto {} => Self::Crypto(),
-            AdditionalPaymentData::BankDebit {} => Self::BankDebit(),
+            AdditionalPaymentData::PayLater {} => Self::PayLater,
+            AdditionalPaymentData::Wallet {} => Self::Wallet,
+            AdditionalPaymentData::BankRedirect { bank_name: _ } => Self::BankRedirect,
```
Comment (SanchithHegde) [inline crates/api_models/src/payments.rs:1802]: Optional change:

```suggestion
            AdditionalPaymentData::BankRedirect { .. } => Self::BankRedirect,
```

File: crates/router/src/core/payments.rs [hunk 10]
```diff
@@ -41,7 +41,7 @@ use crate::{
     utils::{Encode, OptionExt, ValueExt},
 };
 
-#[instrument(skip_all, fields(payment_id, merchant_id))]
+// #[instrument(skip_all, fields(payment_id, merchant_id))]
```
Comment (SanchithHegde) [inline crates/router/src/core/payments.rs:44]: Please revert this change.

File: crates/cards/src/validate.rs (hunk 8 above)
Comment (sahkal) [reply crates/cards/src/validate.rs]: done!

File: crates/api_models/src/payments.rs (hunk 9 above)
Comment (sahkal) [reply crates/api_models/src/payments.rs]: done!

File: crates/router/src/core/payments.rs (hunk 10 above)
Comment (sahkal) [reply crates/router/src/core/payments.rs]: done, reverted!

File: crates/cards/src/validate.rs [hunk 11]
```diff
@@ -22,6 +22,9 @@ impl CardNumber {
     pub fn get_card_isin(self) -> String {
         self.0.peek().chars().take(6).collect::<String>()
     }
+    pub fn get_last4(self) -> String {
+        self.0.peek().chars().rev().take(4).collect::<String>()
```
Comment (SanchithHegde) [inline crates/cards/src/validate.rs:26]: If your last 4 digits are `1234`, this would return `4321`, isn't it?

Adding a `rev()` after the `take()` should provide us the intended result.

```suggestion
        self.0.peek().chars().rev().take(4).rev().collect::<String>()
```

Comment (sahkal) [reply crates/cards/src/validate.rs:26]: self.0.peek().chars().rev().take(4).rev().collect::<String>()
   |                                                   ^^^^^^^ method cannot be called on `Rev<Take<Rev<Chars<'_>>>>` due to unsatisfied trait bounds
   |
   
   i am getting this error, if i try to rev() after take(4)

Comment (SanchithHegde) [reply crates/cards/src/validate.rs:26]: In that case, please revert to your original implementation with the intermediate `collect()`.

Comment (sahkal) [reply crates/cards/src/validate.rs]: done!

**Review by SanchithHegde — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]
